python/ql/src/semmle/python/web/django/Request.qll

```diff
@@ -76,41 +76,18 @@ class DjangoClassBasedViewRequestArgument extends DjangoRequestSource {
     }
 }
 
-/* ***********  Routing  ********* */
-/* Function based views */
-predicate url_dispatch(CallNode call, ControlFlowNode regex, FunctionValue view) {
-    exists(FunctionValue url |
-        Value::named("django.conf.urls.url") = url and
-        url.getArgumentForCall(call, 0) = regex and
-        url.getArgumentForCall(call, 1).pointsTo(view)
-    )
-}
-
-class UrlRegex extends RegexString {
-    UrlRegex() { url_dispatch(_, this.getAFlowNode(), _) }
-}
-
-class UrlRouting extends CallNode {
-    UrlRouting() { url_dispatch(this, _, _) }
-
-    FunctionValue getViewFunction() { url_dispatch(this, _, result) }
-
-    string getNamedArgument() {
-        exists(UrlRegex regex |
-            url_dispatch(this, regex.getAFlowNode(), _) and
-            regex.getGroupName(_, _) = result
-        )
-    }
-}
-
 /** An argument specified in a url routing table */
-class HttpRequestParameter extends HttpRequestTaintSource {
-    HttpRequestParameter() {
-        exists(UrlRouting url |
-            this.(ControlFlowNode).getNode() = url
-                        .getViewFunction()
-                        .getScope()
-                        .getArgByName(url.getNamedArgument())
+class DjangoRequestParameter extends HttpRequestTaintSource {
+    DjangoRequestParameter() {
+        exists(DjangoRoute route, Function f |
+            f = route.getViewFunction().getScope() |
+            this.(ControlFlowNode).getNode() = f.getArgByName(route.getNamedArgument())
+            or
+            exists(int i | i >= 0 |
+                i < route.getNumPositionalArguments() and
+                // +1 because first argument is always the request
+                this.(ControlFlowNode).getNode() = f.getArg(i+1)
```

Comment:
I have not tested this, but perhaps you could rewrite `this.(ControlFlowNode).getNode() = foo` as `this = foo.getAFlowNode()`? (Likewise on line 84.)

Reply:
I can try this out!

Reply:
We discussed this in person. It is possible to write `this = f.getArg(i+1).(AstNode).getAFlowNode()`, but since that was not super elegant, we kept the current solution.